pkg/backoff/backoff_test.go

```diff
@@ -0,0 +1,87 @@
+package backoff_test
+
+import (
+	"errors"
+	"io/ioutil"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/spf13/viper"
+	"github.com/wakatime/wakatime-cli/pkg/backoff"
+	"github.com/wakatime/wakatime-cli/pkg/heartbeat"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestWithRetry(t *testing.T) {
+	v := viper.New()
+
+	tmpFile, err := ioutil.TempFile(os.TempDir(), "wakatime")
+	require.NoError(t, err)
+
+	defer os.Remove(tmpFile.Name())
+
+	v.Set("config", tmpFile.Name())
+
+	opt := backoff.WithRetry(backoff.RetryConfig{
+		V: v,
+	})
+
+	handle := opt(func(hh []heartbeat.Heartbeat) ([]heartbeat.Result, error) {
+		return []heartbeat.Result{
+			{
+				Status: 201,
+			},
+		}, nil
+	})
+
+	_, err = handle([]heartbeat.Heartbeat{})
+	require.NoError(t, err)
+}
+
+func TestWithRetry_BeforeNextBackoff(t *testing.T) {
+	backoffAt := time.Now().Add(time.Second * -1)
+
+	opt := backoff.WithRetry(backoff.RetryConfig{
+		At: &backoffAt,
+	})
+
+	handle := opt(func(hh []heartbeat.Heartbeat) ([]heartbeat.Result, error) {
+		return []heartbeat.Result{
+			{
+				Status: 201,
+			},
+		}, nil
+	})
+
+	_, err := handle([]heartbeat.Heartbeat{})
+	require.Error(t, err)
```

Comment:
Here it could make sense to use a custom error like `backoff.ErrBackoffActive`. This would allow to differentiate this from a general error on sending by previous heartbeat pipeline steps. What do you think?

Reply:
Being honest I haven't decided to create a new custom error because even if an error during backoff occurs the offline step needs to enque the current heartbeat. If you have other thoughts please let me know.

Reply:
yeah, you're right, from the perspective of offline queue, there is no difference actually. :+1: